### PR TITLE
Refactor JsonExpressionBuilder.php

### PR DESCRIPTION
**Pull request by Tigrov**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | -

### CODE REVIEW

**Comment by terabytesoftw** [comment]
Hi, i really don't like the change, why change the interface for the implementation?

**Comment by Tigrov** [comment]
> Hi, i really don't like the change, why change the interface for the implementation?

It is should be `JsonExpression`
https://github.com/yiisoft/db-pgsql/blob/master/src/Builder/JsonExpressionBuilder.php#L69

And `ExpressionInterface` does not have `getValue()` and `getType()` methods

File: src/Builder/JsonExpressionBuilder.php [hunk 1]
```diff
@@ -39,13 +38,11 @@ public function __construct(private QueryBuilderInterface $queryBuilder)
...
-     * @psalm-param JsonExpression $expression
      */
-    public function build(ExpressionInterface $expression, array &$params = []): string
+    public function build(JsonExpression $expression, array &$params = []): string
     {
         /**
-         * @psalm-var array|mixed|QueryInterface $value
+         * @psalm-var array|mixed|QueryInterface|ArrayExpression $value
          */
         $value = $expression->getValue();
```
Comment (vjik) [inline src/Builder/JsonExpressionBuilder.php:47]: `$value` is mixed according to this code: https://github.com/yiisoft/db/blob/master/src/Expression/JsonExpression.php#L32

**Comment by vjik** [comment]
> Hi, i really don't like the change, why change the interface for the implementation?

This PR not change interface. `ExpressionBuilderInterface` is empty: https://github.com/yiisoft/db/blob/master/src/Expression/ExpressionBuilderInterface.php

**Review by terabytesoftw — APPROVED** [review]

**Review by vjik — CHANGES_REQUESTED** [review]
Need changelog only.

File: src/Builder/JsonExpressionBuilder.php [hunk 2]
```diff
@@ -65,8 +63,10 @@ public function build(ExpressionInterface $expression, array &$params = []): str
 
     /**
      * @return string The typecast expression based on {@see JsonExpression::getType()}.
+     *
+     * @psalm-param JsonExpression $expression
      */
-    protected function getTypecast(JsonExpression $expression): string
+    private function getTypecast(ExpressionInterface $expression): string
```
Comment (vjik) [inline src/Builder/JsonExpressionBuilder.php:69]: Keep `JsonExpression`. If check `$expression` in `build()` then here will be `JsonExpression` guaranteed.

File: src/Builder/JsonExpressionBuilder.php [hunk 3]
```diff
@@ -44,9 +44,7 @@ public function __construct(private QueryBuilderInterface $queryBuilder)
      */
     public function build(ExpressionInterface $expression, array &$params = []): string
```
Comment (vjik) [inline src/Builder/JsonExpressionBuilder.php:45]: Need check that $expression is instance of `JsonExpression` here.

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/Builder/JsonExpressionBuilder.php (hunk 3 above)
Comment (vjik) [reply src/Builder/JsonExpressionBuilder.php:43]: See https://github.com/yiisoft/db-pgsql/pull/300#discussion_r1277069846

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 ## 1.1.1 under development
 
-- no changes in this release.
+- Chg #301: Refactor JsonExpressionBuilder (@Tigrov)
```
Comment (vjik) [inline CHANGELOG.md:5]:
```suggestion
- Enh #301: Refactor `JsonExpressionBuilder` (@Tigrov)
```

**Review by vjik — APPROVED** [review]

**Comment by vjik** [comment]
@Tigrov Thanks 👍

**Pull request merged**